Feat controlled custom gate

MathieuG-Colibri opened the pull request:
Added the Custom Controlled Gate class with its to_other_language methods.

Also made CustomGate initialization more flexible for the user.

Henri-ColibrITD committed:
Merge pull request #65 from ColibrITD-SAS/dev

fix: relaxed the versions dependency so the install works on mac

Henri-ColibrITD committed:
Update doc.yml

hJaffaliColibritd committed:
Merge pull request #66 from ColibrITD-SAS/Henri-ColibrITD-patch-2

Update doc.yml

Henri-ColibrITD committed:
Update CONTRIBUTING.md

Henri-ColibrITD committed:
Merge pull request #83 from ColibrITD-SAS/Henri-ColibrITD-patch-2

Update CONTRIBUTING.md

Henri-ColibrITD committed:
Update README.md

Henri-ColibrITD committed:
Merge pull request #96 from ColibrITD-SAS/Henri-ColibrITD-patch-2

Update README.md

Henri-ColibrITD committed:
Update CONTRIBUTING.md

Henri-ColibrITD committed:
Merge pull request #105 from ColibrITD-SAS/Henri-ColibrITD-patch-2

Update CONTRIBUTING.md

Henri-ColibrITD committed:
Merge pull request #113 from ColibrITD-SAS/dev

New release: 0.3

Henri-ColibrITD committed:
Merge pull request #114 from ColibrITD-SAS/dev

pre-version patch

Henri-ColibrITD committed:
Merge pull request #117 from ColibrITD-SAS/dev

Version 0.3.1

hJaffaliColibritd committed:
Merge pull request #124 from ColibrITD-SAS/dev

Version 0.3.2

Henri-ColibrITD committed:
Update pipy.yml

Henri-ColibrITD committed:
Merge pull request #125 from ColibrITD-SAS/Henri-ColibrITD-patch-1

Update pipy.yml

hJaffaliColibritd committed:
Merge pull request #141 from ColibrITD-SAS/dev

Version 0.4.0

MathieuG-Colibri committed:
feat: added CustomControlledGate for Qiskit and QASM2

MathieuG-Colibri committed:
feat: added to_other_language implementation for every languages

MathieuG-Colibri committed:
fix: transpile the circuit if its ran with qiskit with cu gates

MathieuG-Colibri committed:
feat: added documentation and tests

MathieuG-Colibri committed:
fix: python 3.9 tests

MathieuG-Colibri committed:
fix: python3.9 test again

MathieuG-Colibri committed:
fix: deleted unused class

MathieuG-Colibri committed:
fix: deleted useless import

hJaffaliColibritd commented on mpqp/core/instruction/gates/controlled_custom_gate.py:
```diff
@@ -0,0 +1,217 @@
+
+from .gate import Gate
+
+ParametrizedGates = [Rz, Ry, Rx]
+OneQubitGates = [H, X, Y, Z, Id]
```
I would use the predefiined class and a test like `issubclass(gate_class, ParametrizedGate` and same with `SingleQubitGate` or `OneQubitNoParamGate`, because all the native gate have multiple heritage from abstract classes so we can quickly know their properties (rotation or not, parameteized or not, single qubit, controlled, balbalbabl)

MathieuG-Colibri replied:
Fixed in https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/ee0fca19537c45aae1e6efff2ca4e18c0e0dfbc3

With the changed initialization.

hJaffaliColibritd commented on mpqp/core/instruction/gates/controlled_custom_gate.py:
```diff
@@ -0,0 +1,217 @@
+        self,
+        controls: Union[list[int], int],
+        targets: Union[list[int], int],
+        gate: Union[type[Gate], Matrix],
```
I wouldn't allow to put directly a Matrix here, but rather a `UnitaryMatrix`  or even better a `CustomGate`, so you have directly the targets information.

MathieuG-Colibri replied:
Fixed in : https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/ee0fca19537c45aae1e6efff2ca4e18c0e0dfbc3

Now the gate is initialized by the user instead of in the init

hJaffaliColibritd commented on mpqp/core/instruction/gates/controlled_custom_gate.py:
```diff
@@ -0,0 +1,217 @@
+        self,
+        controls: Union[list[int], int],
+        targets: Union[list[int], int],
+        gate: Union[type[Gate], Matrix],
```
I would put a `NativeGate` instead of `type[Gate]` so you directly have an instance which contains the targets of the gate, and you don't need to put it in the arguments.

MathieuG-Colibri replied:
Fixed in https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/ee0fca19537c45aae1e6efff2ca4e18c0e0dfbc3

hJaffaliColibritd commented on mpqp/core/instruction/gates/controlled_custom_gate.py:
```diff
@@ -0,0 +1,217 @@
+        controls: Union[list[int], int],
+        targets: Union[list[int], int],
+        gate: Union[type[Gate], Matrix],
+        rotation: Optional[float] = None,
```
I am not fan of this parameter. If the gate is a parametrized gate, and if we input directly the instance of the gate (see my proposition in the previous comments), then it will directly contain the parameter(s) needed.

ALso here you don't take into account the classes that have multi parameters (Rk, U, ...)

MathieuG-Colibri replied:
Fixed in https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/ee0fca19537c45aae1e6efff2ca4e18c0e0dfbc3

hJaffaliColibritd reviewed (changes requested):
Just started quickly to review, I have to finish

MathieuG-Colibri committed:
fix: made init simpler and more future proof

MathieuG-Colibri committed:
fix: tests

MathieuG-Colibri committed:
fix: deleting trash

MoHermes commented:
@MathieuG-Colibri I changed the base branch of the PR to `dev` instead of `main`,  so we can first review and test the changes in our `dev` environment to ensure stability before merging into main.

MoHermes commented on mpqp/core/circuit.py:
```diff
@@ -1292,7 +1293,9 @@ def to_other_language(
                 if isinstance(instruction, (ExpectationMeasure, Barrier, Breakpoint)):
                     continue
-                elif isinstance(instruction, CustomGate):
+                elif isinstance(instruction, CustomGate) or isinstance(
+                    instruction, CustomControlledGate
+                ):
```
```suggestion
                elif isinstance(instruction, (CustomGate, CustomControlledGate)):
```

MathieuG-Colibri replied:
Refactored : https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/14ebd270f61c4e47baaf245af5523f86bc7b7721

Note : the filename has been changed.

MoHermes commented on mpqp/core/instruction/gates/controlled_custom_gate.py:
```diff
@@ -0,0 +1,145 @@
+        else:
+            ControlledGate.__init__(self, controls, gate.targets, gate, label)
+
+    def __repr__(self) -> str:
```
Here, it needs to be refactored a bit for clarity, for example: `representation` repeatedly mutated!!, formatting `controls` and `targets` in a way that helps,  also we could just have one return!

MathieuG-Colibri replied:
Refactored : https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/14ebd270f61c4e47baaf245af5523f86bc7b7721

Note : the filename has been changed.

MathieuG-Colibri committed:
fix: refactor repr method and changed filename

JulienCalistoTD commented on mpqp/execution/providers/ibm.py:
```diff
@@ -454,6 +454,14 @@ def run_aer(job: Job):
+        if any(isinstance(gate, CustomControlledGate) for gate in job.circuit.gates):
+            from qiskit import transpile
+
+            print("pog pog pog pog pog pog")
```
POG POG POG !

MathieuG-Colibri replied:
Removed https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

JulienCalistoTD commented on mpqp/execution/providers/ibm.py:
```diff
@@ -454,6 +454,14 @@ def run_aer(job: Job):
+
+            print("pog pog pog pog pog pog")
+            assert isinstance(qiskit_circuit, QuantumCircuit)
+            qiskit_circuit = transpile(qiskit_circuit, AerSimulator())
```
Don't transpile the circuit here, check `to_other_device` in QCircuit

MathieuG-Colibri replied:
Moved https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

JulienCalistoTD commented on mpqp/qasm/mpqp_to_qasm.py:
```diff
@@ -70,7 +72,9 @@ def _simplify_instruction_to_qasm(
     if isinstance(instruction, (Breakpoint, ExpectationMeasure)):
         return "", 0
-    elif isinstance(instruction, CustomGate):
+    elif isinstance(instruction, CustomGate) or isinstance(
```
```suggestion
    elif isinstance(instruction, (CustomGate, CustomControlledGate)):
```

MathieuG-Colibri replied:
Done https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

JulienCalistoTD commented on mpqp/core/instruction/gates/custom_controlled__gate.py:
```diff
@@ -0,0 +1,140 @@
+        else:
+            ControlledGate.__init__(self, controls, gate.targets, gate, label)
+
+    def __repr__(self) -> str:
```
a repr should be a str to rebuild the object

MathieuG-Colibri replied:
Fixed ! https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

MoHermes commented on mpqp/core/instruction/gates/controlled_gate.py:
```diff
@@ -25,11 +25,15 @@ class ControlledGate(Gate, ABC):
         self,
-        controls: list[int],
-        targets: list[int],
+        controls: Union[list[int], int],
+        targets: Union[list[int], int],
```
I prefer to use a reusable type alias like for ex. `QubitIndex = Union[int, list[int]]` as it is used multiple times in the code.

MathieuG-Colibri replied:
I agree but this specific union is used once or twice per file where it is used. in that case isn't it more clear for the user to have this simple Union ?

Also do we want to have even more small imports like this in the project ?

Henri-ColibrITD replied:
also, the doc generator doesn't handle aliases super smoothly

MoHermes replied:
Ok , we keep it simple in this case for the minimal usage and docgen issues .

MoHermes commented on mpqp/execution/providers/ibm.py:
```diff
@@ -454,6 +454,14 @@ def run_aer(job: Job):
+        if any(isinstance(gate, CustomControlledGate) for gate in job.circuit.gates):
+            from qiskit import transpile
+
+            print("pog pog pog pog pog pog")
```
Don't forget to remove it!

MathieuG-Colibri replied:
Removed ! https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

MoHermes commented on mpqp/core/instruction/gates/custom_controlled__gate.py:
Why double underscore, intentional reason?

MathieuG-Colibri replied:
It wasn't thanks for pointing it out https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/a6827c9b907368956da9319565a99103bdce4877

MathieuG-Colibri committed:
fix: repr fix and negative indices error

Henri-ColibrITD commented on mpqp/core/circuit.py:
```diff
@@ -1292,7 +1293,7 @@ def to_other_language(
                 if isinstance(instruction, (ExpectationMeasure, Barrier, Breakpoint)):
                     continue
-                elif isinstance(instruction, CustomGate):
+                elif isinstance(instruction, (CustomGate, CustomControlledGate)):
```
if customcontrolledGate inherits from customgate, you don't need to add this check, and if not, should it ?

MathieuG-Colibri replied:
I don't like it because the total number of targets of these types of gates don't match with CustomGate's.

It's just sometimes it's more useful to be associated with a ControlledGate and sometimes it can be treated as a CustomGate.

Also CustomGate adds some "useless" properties to this.

Henri-ColibrITD commented on mpqp/core/circuit.py:
```diff
@@ -1493,6 +1494,14 @@ def to_other_device(
+                    assert isinstance(qiskit_circuit, QuantumCircuit)
+                qiskit_circuit = transpile(qiskit_circuit, backend_sim)
             return qiskit_circuit
         elif isinstance(device, GOOGLEDevice):
```
is this new type of gate supported for all providers ?

MathieuG-Colibri replied:
Yes when translated through QASM this gate is decomposed it's only in qiskit where it is compatible, although it need to be transpiled to be ran.

Henri-ColibrITD commented on mpqp/core/instruction/gates/controlled_gate.py:
```diff
@@ -25,17 +25,25 @@ class ControlledGate(Gate, ABC):
         self,
-        controls: list[int],
-        targets: list[int],
+        controls: Union[list[int], int],
```
is this change really needed ?

MathieuG-Colibri replied:
I like it for flexibility reasons, I don't like having to call [0] when having one control.

Same reason when having to call another constructor inside our constructor like CustomGate(UnitaryMatrix( I think it could be improved.

Henri-ColibrITD commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,122 @@
+            if isinstance(controls, int):
+                controls = [controls]
+            controls += gate.controls
+            ControlledGate.__init__(
```
you call the `__init__` for a specific reason ?

Henri-ColibrITD commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,122 @@
+    def __repr__(self) -> str:
+        return f"CustomControlledGate({self.controls}, {self.non_controlled_gate.__repr__()})"
+
+    def to_canonical_matrix(self):
```
it's been a long time so I don't remember precisely but the to matrix is implemented in the parent class ? no need to implement it here ?

MathieuG-Colibri replied:
Yes it is implemented in ControlledGate already.

Henri-ColibrITD commented on mpqp/core/instruction/gates/custom_gate.py:
```diff
@@ -50,11 +50,17 @@ class CustomGate(Gate):
     def __init__(
-        self, definition: UnitaryMatrix, targets: list[int], label: Optional[str] = None
+        self,
+        definition: Union[UnitaryMatrix, Matrix],
```
should we really have a plain matrix here ? the reason why we imposed unitarymatrix here was to enforce the unitarity check. If you think this is not userfriendly, maybe then we should change how this check is performed, but I'm not fan of the union

MathieuG-Colibri replied:
I think so that we can remove the mandatory UnitaryMatrix call everytime we call CustomGate.

Otherwise we could just set definition as Matrix and force the user to give the matrix we're going to put through UnitaryMatrix, but I think leaving the choice up to the user is best of both worlds.

Henri-ColibrITD replied:
I say we move to 100% matrix

MathieuG-Colibri replied:
Moved to matrix in [fed43e3](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/fed43e3506d4da71e988dcd0e4f545bb353d2bb1)

Henri-ColibrITD commented on mpqp/core/instruction/gates/custom_gate.py:
```diff
@@ -50,11 +50,17 @@ class CustomGate(Gate):
+            targets = [targets]
         self.definition = definition
         """See parameter description."""
-
```
reason for these deletions ?

MathieuG-Colibri replied:
My theory it's because of the formatter that changed when I added code then deleted it.

Henri-ColibrITD commented on mpqp/core/instruction/gates/custom_gate.py:
```diff
@@ -154,14 +160,13 @@ def to_other_language(
     def __repr__(self) -> str:
         label = f", \"{self.label}\"" if self.label else ""
-        return f"CustomGate({UnitaryMatrix(self.matrix)}, {self.targets}{label})"
+        return f"CustomGate({UnitaryMatrix(self.matrix).__repr__()}, {self.targets}{label})"
```
```suggestion
        return f"CustomGate({repr(UnitaryMatrix(self.matrix))}, {self.targets}{label})"
```

Henri-ColibrITD commented on mpqp/core/instruction/gates/gate_definition.py:
```diff
@@ -85,7 +85,7 @@ def inverse(self) -> GateDefinition:
         Example:
             >>> UnitaryMatrix(np.array([[1, 0], [0, -1]])).inverse()
-            UnitaryMatrix(array([[ 1., 0.], [-0., -1.]]))
+            UnitaryMatrix(np.array([[ 1., 0.], [-0., -1.]]))
```
I prefer the way it was before: more standard

Henri-ColibrITD commented on mpqp/core/instruction/gates/gate_definition.py:
```diff
@@ -255,4 +255,4 @@ def caster(v: Expr | Complex) -> Expr | Complex:
 
     def __repr__(self) -> str:
-        return f"UnitaryMatrix({one_lined_repr(self.matrix)})"
+        return f"UnitaryMatrix(np.{one_lined_repr(self.matrix)})"
```
see above

Henri-ColibrITD commented on mpqp/qasm/mpqp_to_qasm.py:
```diff
@@ -70,7 +72,7 @@ def _simplify_instruction_to_qasm(
     if isinstance(instruction, (Breakpoint, ExpectationMeasure)):
         return "", 0
-    elif isinstance(instruction, CustomGate):
+    elif isinstance(instruction, (CustomGate, CustomControlledGate)):
```
see comment inheritance comment above

Henri-ColibrITD commented on mpqp/qasm/open_qasm_2_and_3.py:
```diff
@@ -362,7 +362,10 @@ def add_std_lib():
     elif instr_name == "gate":
-        defined_gates.add(instr.split()[1])
+        g_name = instr.split()[1]
+        if g_name.count('(') != 0:
```
I think this test is not even needed: if you have no parenthesis, `g_name.split('(')[0]` is the identity

MathieuG-Colibri replied:
It is used to extract the name of a parametrized gate which are recognizable by having parenthesis.

It's to check if when calling Rx(pi) that Rx does exist.

Henri-ColibrITD commented on tests/core/instruction/gates/test_custom_controlled_gate.py:
```diff
@@ -0,0 +1,31 @@
+    assert gate == eval(gate.__repr__())
+
+
+def test_negative_indices():
+    with pytest.raises(ValueError):
+        CNOT(-1, 0)
+    with pytest.raises(ValueError):
+        TOF([1, -1], 0)
+    with pytest.raises(ValueError):
+        CustomControlledGate(-10, Z(0))
+    with pytest.raises(ValueError):
+        CustomControlledGate([0, 1, -3], S(2))
```
```suggestion
@pytest.mark.parametrize(
    "gate, args",
    [
        (CNOT, (-1, 0)),
        (TOF, ([1, -1], 0)),
        (CustomControlledGate, (-10, Z(0))),
        (CustomControlledGate, ([0, 1, -3], S(2))),
    ],
)
def test_negative_indices(gate: Gate, args: tuple[Any]):
    with pytest.raises(ValueError):
        gate(*args)
```

Henri-ColibrITD reviewed (changes requested):
a few changes to justify

MathieuG-Colibri committed:
Update tests/core/instruction/gates/test_custom_controlled_gate.py

Co-authored-by: Henri de Boutray <133855265+Henri-ColibrITD@users.noreply.github.com>

MathieuG-Colibri committed:
Update mpqp/core/instruction/gates/custom_gate.py

Co-authored-by: Henri de Boutray <133855265+Henri-ColibrITD@users.noreply.github.com>

Commit:
test: add test scenarios for gates with parameters

Custom gates with parameters cause an exception if tried to compile them from OpenQASM 2.0 to 3.0.
Likewise, gates with parameters in an if-body cause an exception.

MathieuG-Colibri committed:
fix: tests

MathieuG-Colibri reviewed (commented):
see pending comments

Henri-ColibrITD committed:
doc: details

Henri-ColibrITD commented on mpqp/qasm/open_qasm_2_and_3.py:
```diff
@@ -362,7 +362,10 @@ def add_std_lib():
-        defined_gates.add(instr.split()[1])
+        g_name = instr.split()[1]
+        if g_name.count('(') != 0:
+            g_name = g_name.split('(')[0]
```
```suggestion
        g_name = g_name.split('(')[0]
```

Henri-ColibrITD committed:
Update mpqp/qasm/open_qasm_2_and_3.py

MathieuG-Colibri committed:
fix: readded the np to repr

MathieuG-Colibri committed:
Merge branch 'feat-controlled-custom-gate' of github.com:ColibrITD-SAS/mpqp into feat-controlled-custom-gate

Henri-ColibrITD committed:
Merge pull request #155 from LiRem101/test-gates-with-params

Tests for gates with parameters

github-actions[bot] committed:
chore: Files formated

MathieuG-Colibri committed:
fix: changed definition from UnitaryMatrix to Matrix type

MathieuG-Colibri committed:
Merge branch 'feat-controlled-custom-gate' of github.com:ColibrITD-SAS/mpqp into feat-controlled-custom-gate

MathieuG-Colibri committed:
fix: tests

Henri-ColibrITD commented on mpqp/core/instruction/gates/gate_definition.py:
```diff
@@ -158,6 +157,12 @@ class UnitaryMatrix(GateDefinition):
+        ValueError: Matrices defining gates have to be unitary.
+
+        ValueError: The unitary matrix of a gate acting on qubits must have
+        dimensions that are power of two.
```
format issue
```suggestion
        ValueError: Matrices defining gates have to be unitary.
        ValueError: The unitary matrix of a gate acting on qubits must have
            dimensions that are power of two.
```

MathieuG-Colibri committed:
Update mpqp/core/instruction/gates/gate_definition.py

Co-authored-by: Henri de Boutray <133855265+Henri-ColibrITD@users.noreply.github.com>

JulienCalistoTD commented on mpqp/core/circuit.py:
```diff
@@ -1493,6 +1493,14 @@ def to_other_device(
                 backend = get_backend(device)
                 pm = generate_preset_pass_manager(backend=backend, optimization_level=1)
                 qiskit_circuit = pm.run(qiskit_circuit)
+            if any(
```
```suggestion
            # TODO: removed with PR - Circuit Handling and PauliString Utilities #154
            if any(
```

JulienCalistoTD commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+    def __repr__(self) -> str:
+        return f"CustomControlledGate({self.controls}, {self.non_controlled_gate.__repr__()})"
+
+    def to_matrix(self, desired_gate_size: int = 0) -> Matrix:
```
`to_matrix` is alredy define in `ControledGate` so you don't need to overwrite this function

Henri-ColibrITD replied:
done in [7a80438](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/7a80438e5a2d90129611691b1f17873c11da642f)

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+        controls: List of indices referring to the qubits used to control the gate.
+        targets: List of indices referring to the qubits on which the gate will be applied.
+        non_controlled_gate: The original, non controlled, gate or a matrix.
+        label: Label used to identify the gate.
```
Update Args to match the new argumetns of the constrcutor

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+class CustomControlledGate(ControlledGate):
+    """
+    Class used to define a custom controlled gate.
+    It can be either a native gate with any numbers of control qubits or a custom gate with control qubits.
```
```suggestion
    It can be either a native gate or a CustomGate, with any number of control qubits.
```

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+    q_1: ┤ Unitary ├
+         └────┬────┘
+    q_2: ─────■─────
+
```
add indentation of a tabulation with respect to the title `Examples:`

MathieuG-Colibri replied:
Done [7a80438](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/7a80438e5a2d90129611691b1f17873c11da642f)

hJaffaliColibritd commented on mpqp/core/circuit.py:
```diff
@@ -848,7 +848,7 @@ def to_gate(self) -> Gate:
         # 3M-TODO check implementation, example and test, this will only work
            when circuit.to_matrix() will be implemented
```
can we remove this TODO ?

hJaffaliColibritd commented on tests/core/instruction/gates/test_custom_controlled_gate.py:
```diff
@@ -0,0 +1,42 @@
+)
+def test_negative_indices(gate: Type[Gate], args: tuple[Any]):
+    with pytest.raises(ValueError):
+        gate(*args)
```
An example that made it crash : 

```python
from mpqp.all import *
c = QCircuit([CustomControlledGate([1],CNOT(0,2))])
c.inverse()
```

The trace : 
```python
ValueError                                Traceback (most recent call last)
<ipython-input-13-ca58f1e38c27> in <module>
----> 1 c.inverse()

~\mpqp\core\circuit.py in inverse(self)
    825         for instr in self.instructions:
    826             if isinstance(instr, Gate):
--> 827                 dagger.instructions.insert(0, instr.inverse())
    828             elif isinstance(instr, Barrier):
    829                 dagger.instructions.insert(0, instr)

~\mpqp\core\instruction\gates\gate.py in inverse(self)
    162         from mpqp.core.instruction.gates.custom_gate import CustomGate
    163
--> 164         return CustomGate(
    165             self.to_matrix().transpose().conjugate(),
    166             self.targets,

~\mpqp\core\instruction\gates\custom_gate.py in __init__(self, matrix, targets, label)
     61         """See parameter description."""
     62         if definition.nb_qubits != len(targets):
---> 63             raise ValueError(
     64                 f"Size of the targets ({len(targets)}) must match the number of qubits of the "
     65                 f"UnitaryMatrix ({definition.nb_qubits})"

ValueError: Size of the targets (1) must match the number of qubits of the UnitaryMatrix (3)
```

I think it comes from the fact that we don't update correctly the `definition` of the gate when controlled

MathieuG-Colibri replied:
Done ! [dc7e540](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/dc7e5401d008b616113e02b0f6c75227c910136e)

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+from typing import Any, Optional, Union
+from qiskit.circuit import Parameter
```
Import it at the place in the code when needed, not here

MathieuG-Colibri replied:
Fixed (put in type checking) [7a80438](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/7a80438e5a2d90129611691b1f17873c11da642f)

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+        if desired_gate_size != 0:
+            max_qubit = max(control, target) + 1
+            if desired_gate_size < max_qubit:
+                raise ValueError(f"nb_qubits must be at least {max_qubit}")
```
Message not clear. Nb_qubits of what ?

MathieuG-Colibri replied:
Fixed [62414f9](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/62414f9a551d6e5258e7952b7d2c0b63a38edec9)

hJaffaliColibritd commented on tests/core/instruction/gates/test_custom_gate.py:
```diff
@@ -4,7 +4,7 @@
 import numpy as np
 import pytest
-
+from numpy import array  # pyright: ignore[reportUnusedImport]
```
if it is non used then just remove the import , no ?

MathieuG-Colibri replied:
Used for repr, by design numpy output array when we use repr on it so we need to import it that way.

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+                controls = [x - min_qubit for x in controls]
+                targets = [x - min_qubit for x in targets]
+            elif desired_gate_size < max_qubit + 1:
+                raise ValueError(f"nb_qubits must be at least {max_qubit + 1}")
```
Error message not clear, nb_qubits of what ?

MathieuG-Colibri replied:
Fixed [62414f9](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/62414f9a551d6e5258e7952b7d2c0b63a38edec9)

Henri-ColibrITD replied:
actually 7a80438

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+        if language == Language.QISKIT:
+            from qiskit.quantum_info import Operator
+
+            gate = self.non_controlled_gate.to_other_language()
```
what is it is not coming from a native gate ? it will control the `unitary` custom gate ?

MathieuG-Colibri replied:
Can you rephrase please sir ?

hJaffaliColibritd replied:
sorry I wanted to say "what if it is not coming from NativeGate" meaning that the controlled gate can come from a custom gate right ?

hJaffaliColibritd replied:
I double checked I think I have my answer, all good.

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_controlled_gate.py:
```diff
@@ -0,0 +1,194 @@
+
+            if isinstance(self.non_controlled_gate, CustomGate):
+                targets = self.targets + self.controls
+                targets.sort()
```
Can this f*ck up the non ordered targets ?

MathieuG-Colibri replied:
No although it may cause some problems with non continuous targets.
But it's handled by qiskit.

hJaffaliColibritd replied:
Ok

hJaffaliColibritd commented on mpqp/core/instruction/gates/custom_gate.py:
```diff
@@ -154,14 +158,13 @@ def to_other_language(
     def __repr__(self) -> str:
         label = f", \"{self.label}\"" if self.label else ""
-        return f"CustomGate({UnitaryMatrix(self.matrix)}, {self.targets}{label})"
+        return f"CustomGate({repr(self.matrix)}, {self.targets}{label})"
```
this can introduce problems with the repr of the matrix which is a numpy array

```python
from mpqp.all import *
cu = CustomGate(np.array([[1,0],[0,1]]), 0)
c2 = eval(repr(cu))
```

Giving the error 

```python
NameError: name 'array' is not defined
```

MathieuG-Colibri replied:
That is normal it's how numpy does it.
Henri wants repr to be this way and I agree

hJaffaliColibritd commented on tests/core/instruction/gates/test_custom_controlled_gate.py:
```diff
@@ -0,0 +1,42 @@
+from typing import Any, Type
```
Add a test for to other language on a circuit containing custom controlled gate , 
and inverse of the circuit/customcontrolledGate ?

MathieuG-Colibri replied:
Done ! [dc7e540](https://github.com/ColibrITD-SAS/mpqp/pull/146/commits/dc7e5401d008b616113e02b0f6c75227c910136e)

MathieuG-Colibri committed:
fix: inverse method and general fixes

MathieuG-Colibri committed:
Merge branch 'feat-controlled-custom-gate' of github.com:ColibrITD-SAS/mpqp into feat-controlled-custom-gate

Henri-ColibrITD committed:
Merge branch 'feat-controlled-custom-gate' of github.com:ColibrITD-SAS/mpqp into feat-controlled-custom-gate

MathieuG-Colibri committed:
feat: added inverse test

Henri-ColibrITD committed:
Update mpqp/core/circuit.py

Co-authored-by: JulienCalistoTD <164006610+JulienCalistoTD@users.noreply.github.com>

Henri-ColibrITD committed:
chore: review comments

Henri-ColibrITD committed:
Merge branch 'feat-controlled-custom-gate' of github.com:ColibrITD-SAS/mpqp into feat-controlled-custom-gate

MathieuG-Colibri committed:
fix: tests

MathieuG-Colibri committed:
fix: merge
